Made some changes and wanted to see your feedback

s0i opened the pull request:
Hello! First time doing a pull request for an open source project and I wanted to get your feedback. I think the biggest thing is that I implemented Permissions for the analytics dashboard.

s0i committed:
Update to use sessions backend

s0i committed:
Updating middleware

Cut out the threading and the fat in the middleware

- Checking for DNT in META in middleware instead of in static function call in model

s0i committed:
Updating some more

- Updated views to class-based
- Added a PageView manager to help declutter the view class

s0i committed:
Changing form action

- Removed the remnants of the old view
- Simplified default kwargs for filtering but it's still not good enough
- Removed unneeded util module

s0i committed:
New template

- Updating template to be a bit cleaner and straightforward

- Keeping bulma css

s0i committed:
Humanizing stats

s0i committed:
Cleaning up database migrations

s0i committed:
Cleaning up

s0i committed:
Delete manage.py

s0i committed:
Delete __init__.py

s0i committed:
Delete settings.py

s0i committed:
Delete urls.py

s0i committed:
Updating migrations

- Removing test environment

s0i committed:
Merge branch 'master' of https://github.com/s0i/django-privacy-analytics

s0i committed:
Delete wsgi.py

s0i committed:
Update README.md

s0i committed:
Trying ideas for Analytics view

- Setting default filter for queryset as all
- use form to GET time filter
- filter queryset and return

s0i committed:
Updating dashboard

- Added messages for clear confirmation
- Tidied up managers.py

s0i committed:
Update models.py

milesmcc commented:
Wow, thanks for putting this together. Seems like a pretty big refactoring and improvement suite, so I’ll spend some time over the next few days looking it over.

Thanks!

s0i commented:
Awesome! I think the date range filtering can be done a more django-ish way but I couldn’t think of anything

Cheers,

Harley

> On May 10, 2020, at 18:48, R. Miles McCain <notifications@github.com> wrote:
> 
> ﻿
> Wow, thanks for putting this together. Seems like a pretty big refactoring and improvement suite, so I’ll spend some time over the next few days looking it over.
> 
> Thanks!
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.

milesmcc commented on README.md:
```diff
@@ -15,10 +15,10 @@ Django Privacy Analytics fully respects the 'Do Not Track' (DNT) header; if a re
 ### Access
 
-Currently, the dashboard is only visible to superusers. This will be configurable in a future version.
+Analyics dashboard can be controlled by user permissions with the `PermissionsRequiredMixin`
```
Hmm, this is a bit unclear. Can we write out the exact permissions that Django Privacy Analytics uses to determine access control here? `PermissionsRequiredMixin` is just a way to check permissions, after all—can we list the permissions we're checking themselves?

s0i replied:
Right now it uses the basic permissions that Django automatically populates once we import the models, e.g., private_analytics.view_pageview, private_analytics.create_pageview, etc. Giving the option for them to implement more specific permissions which can then be updated via settings which I want to add

milesmcc commented on privacy_analytics/admin.py:
```diff
@@ -1,6 +1,6 @@
 from django.contrib import admin
 from .models import PageView
 
+@admin.register(PageView)
```
Excellent improvement, thanks. Much more idiomatic than what we had before.

milesmcc commented on privacy_analytics/managers.py:
```diff
@@ -0,0 +1,25 @@
+        return self.filter(**kwargs).values('session_key').count()
+
+    def percent_authenticated(self, **kwargs):
+        return self.filter(**kwargs).aggregate(total=(Cast(Count('is_authenticated', filter=Q(is_authenticated=True)), FloatField()) / Cast(Count('pk'), FloatField())) * 100.00)
```
What happens when the data is empty? Wouldn't this error due to dividing by zero, or does the database handle this differently?

s0i replied:
I have updated all of the manager logic to combat this and this was a janky fix

milesmcc commented on privacy_analytics/managers.py:
```diff
@@ -0,0 +1,25 @@
+        return self.filter(**kwargs).aggregate(total=(Cast(Count('is_authenticated', filter=Q(is_authenticated=True)), FloatField()) / Cast(Count('pk'), FloatField())) * 100.00)
+    
+    def referrers(self, **kwargs):
+        return self.filter(**kwargs).values('referrer').annotate(total=Count('pk')).all()[:10]
```
* Does this include the referrers for _all_ page loads, or just initial page loads? I don't remember this codebase well enough to remember if internal referrer handling is handled elsewhere.
* Does this return the top 10 referrers, or does it use some other ordering? Asking mostly because I don't see any sort specified in the db lookup.
* Should we consider adding an option to list more than 10 referrers, maybe as an optional argument? Alternatively, should we just return all of the referrers, and do the slicing elsewhere (in the view, for example)?

s0i replied:
So I have refactored the `managers.py` and have tidied it up. If there are zero records when it does the math, Django handles that internally and creates no issues.

Right now, on the initial page load for the dashboard, it just loads all records from the database. If we implement pagination, which Django supports, we definitely could add in adding the option to limit!

milesmcc commented on privacy_analytics/managers.py:
```diff
@@ -0,0 +1,25 @@
+    def pages(self, **kwargs):
+        return self.filter(**kwargs).values('path').annotate(total=Count('pk'), percent=(Count('pk') * 100) / self.filter(**kwargs).all().count()).order_by('-total')
+    
+    def clear_by_time(self, **kwargs):
```
What's the difference between calling `clear_by_time(...)` and `filter(...).delete()` (where `...` is the same)? Think I might be missing something here — what is the value added here?

s0i replied:
So with this, I pass in kwargs from the forms in the DashboardView, they are in the form of `time__lte` or `time__gte`, etc., (thinking about adding customising options here for more complex usage).

So we get all the PageView objects filtered by our time constraints (if any) and then remove based off of that.

milesmcc commented on privacy_analytics/middleware.py:
```diff
@@ -1,26 +1,21 @@
 class AnalyticsMiddleware:
     def __init__(self, get_response):
         self.get_response = get_response
-        self.processing_queue = queue.Queue()
-        self.processing_thread = threading.Thread(target=self._thread_run)
-        self.processing_thread.daemon = True
-        self.processing_thread.start()
-
-    def _thread_run(self):
-        while True:
-           PageView.create_for_request(self.processing_queue.get())
-           # Will lock until item is available
 
     def __call__(self, request):
         response = self.get_response(request)
-        if hasattr(settings, "ANALYTICS_IGNORE_PATHS"):
-            for path in settings.ANALYTICS_IGNORE_PATHS:
-                if request.path.startswith(path):
-                    return response
-        self.processing_queue.put_nowait(request)
```
Thank you for removing my very janky background queue system. That said, I'm a bit concerned about doing a db write on every page load. Seems like the sort of thing that could overload the server quite easily, as well as be vulnerable to a denial of service attack.

Would you consider moving the write system into a [Celery background task](http://www.celeryproject.org/)?

s0i replied:
Yes! I am going to do so, that is definitely next on the to-do list.

s0i replied:
I have added initial celery support; however, I would like to implement optional support for different task managers, but I am unsure on how to handle that as I've never done something like that before. Could you look into that?

milesmcc commented on privacy_analytics/templates/privacy_analytics/dashboard.html:
```diff
@@ -1,113 +1,149 @@
-    <button type="submit" class="button is-danger is-outlined">Clear Data Before {{from}}</button>
-</form>
+</section>
 {% endblock %}
```
This looks excellent, thanks for building all this. I haven't taken a look at the new interface in the browser, but I'll do so as soon as we move forward with the backend changes!

s0i replied:
It's pretty much a carbon-copy but I just wanted to neaten things up. I am thinking about doing a more intense dashboard mock-up. I will definitely get your thoughts once I whip something up!

milesmcc reviewed (changes requested):
Well done! This is really a wonderful set of improvements, thank you.

I've made a few comments on small changes I think we should consider. Otherwise, though, I'm ready to merge ASAP.

s0i committed:
Refactoring manager logic

- Added a templtae filter for displaying percents
- Temporary workaround for get percent of total in object queries
- Handling clearing of analytics properly (hacky right now)

s0i committed:
Handling empty analytics objects

s0i commented:
@milesmcc I have made some changes to some stuff specifically the handling inside managers.py and preventing some issues that you've mentioned

s0i committed:
Celery support

- Added initial support for Celery, would like to incorporate interchangability with other task handlers

s0i committed:
Updated handling of requests through celery

s0i commented:
I'm not sure if my new commits show up here or not but definitely check out my fork if the changes I remarked are not reflected here

s0i committed:
Fixing issues with SuspiciousOperation error

Trying to track pageviews when user was logged in and then logs out and session changes before processing the request.

milesmcc commented:
Sounds good, and thanks for the ping. It looks like your new commit is showing up here properly. Will take a look ASAP—I'm busy with a bit of work right now, but this is first thing in my review queue!

milesmcc commented:
Looking over this now. Out of curiosity, any reason why you removed the test environment? I agree that it looks a bit weird being in the repository, but it's also an important way to test the library. How did you do your testing/what workflow do you recommend?

s0i commented:
I have a local repo and in my own test project I used pip3 install -e /path/to/repo

Cheers,

Harley

> On May 16, 2020, at 15:03, R. Miles McCain <notifications@github.com> wrote:
> 
> ﻿
> Looking over this now. Out of curiosity, any reason why you removed the test environment? I agree that it looks a bit weird being in the repository, but it's also an important way to test the library. How did you do your testing/what workflow do you recommend?
> 
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub, or unsubscribe.

milesmcc commented:
Alright, I made a few small changes (see [this commit](https://github.com/milesmcc/django-privacy-analytics/commit/a1c50049da662460bef4ae1f8c81b12c60b5e676)), then merged everything in!

Thank you for your additions; these are great. Most of my attention is going towards [Shynet](https://github.com/milesmcc/shynet) right now, but eventually I'd like for django-privacy-analytics to be just as well-featured. This is one step towards that!
